### ISSUE TITLE
[5.1] Eloquent memory issue with eager loading polymorphic relations in v5.1.25

### ISSUE THREAD

**Issue by dwightwatson** [laravel/framework#11138]
The latest release of Laravel has affected a query in one of my apps. The query worked fine in v5.1.24 but now runs out of memory in v5.1.25.

    PHP Fatal error:  Allowed memory size of 536870912 bytes exhausted (tried to allocate 32 bytes) in /home/vagrant/studentvip/vendor/laravel/framework/src/Illuminate/Database/Connection.php on line 323

This is a query that we use to get the message thread between two users, each message in the thread also has polymorphic `messageable` relation

I've paraphrased the code below that shows how this scope works, and how it is used in a controller. Note that `messageble` is eager loaded.

```
class Message extends \Illuminate\Database\Eloquent\Model
{
    public function scopeThread($query, $currentUserId, $otherUserId)
    {
        return $query->with('sender', 'recipient', 'messageable')
            ->where(function ($query) use ($currentUserId, $otherUserId) {
                $query->where('sender_id', $currentUserId)
                    ->where('recipient_id', $otherUserId)
                    ->whereNull('sender_deleted_at');
            })->orWhere(function ($query) use ($currentUserId, $otherUserId) {
                $query->where('sender_id', $otherUserId)
                    ->where('recipient_id', $currentUserId)
                    ->whereNull('recipient_deleted_at');
            })->orderBy('created_at', 'asc');
    }

    public function sender()
    {
        return $this->belongsTo(User::class, 'sender_id');
    }

    public function recipient()
    {
        return $this->belongsTo(User::class, 'recipient_id');
    }

    public function messageable()
    {
        return $this->morphTo();
    }
}
```

If I remove the eager loading of `messageable` from the thread scope then the memory issue is avoided.

I'm not positive, but I suspect [this change to `MorphTo.php`](https://github.com/laravel/framework/commit/aa580c5b4c50127c9e7842c6838afb36a3bc4454#diff-4c052acf0022213a4cc23f53ba42720e) may be the culprit.

**Comment by rkgrep** [comment]
Are you sure it worked before mentioned change with same data?

**Comment by michaelstivala** [comment]
I'm experiencing the same issue with my morphable relations - rolling back to 5.1.24 fixes it.

**Comment by dwightwatson** [comment]
Yeah, definitely worked before. The query above was used on one of our highest trafficked pages and our Sentry log went nuts right after deployment. I suspect similar queries that eager-loaded polymorphic relations would have failed also.

I found that manually reverting the `MorphTo.php` commit I mentioned in the original post fixes this issue, so that's where this problem lies.

**Comment by GrahamCampbell** [comment]
Fixed in v5.1.26, ;)

**Comment by xiphiaz** [comment]
@dwightwatson I'm not sure how the commit can cause a memory issue, as it just moved the logic from `getEager` to `match`, which is where all the other relationships handle their eager loaded model association with parent. 

The commit was to address an inconsistency bug with the relations keying, so a reversion of the commit is not a good idea (reversion would now be a breaking change).

**Comment by xiphiaz** [comment]
@GrahamCampbell, soo...5.1.26 is a breaking change.

**Comment by GrahamCampbell** [comment]
> soo...5.1.26 is a breaking change.

No it's not. That that 5.1.25. 5.1.26 fixes the breaking change.

**Comment by xiphiaz** [comment]
@GrahamCampbell can you explain how 5.1.26 is not a breaking change? It reverts #10864, which fixed #10501. With the reversion, #10501 is re-introduced.

What this means is that developer's morphTo relations will be eager loaded into the model under a snake_cased key, instead of camelCased, which is the pattern all other relations follow. 

I grant that it seems there is a memory issue with the eager loading of morphTo relations that has been introduced by the fix in #10864, but we should be focusing on that rather than swapping one bug for another.

**Comment by GrahamCampbell** [comment]
It's not a breaking change from 5.1.24.

**Comment by GrahamCampbell** [comment]
Consider 5.1.25 to be a broken release.

**Comment by xiphiaz** [comment]
Okay, sure we can go with that. So can you re-open #10501, and I'll re-issue a PR to replicate  #10864, with the intention to resolve this memory issue and adjust the PR to remedy the new bug?

**Comment by GrahamCampbell** [comment]
One open issue is fine I think.

**Comment by xiphiaz** [comment]
Ok that's fine. As you'll see above I've created a new PR to revert the revert - I'm not trying to yo-yo here and create a revert war - this isn't wikipedia ;) I'd like to get this memory issue resolved with the help of @dwightwatson who encountered it, then we can have the best of both worlds - consistent api and no memory issues.

**Comment by dwightwatson** [comment]
Thanks for the quick response shipping a new release.

More than happy to help out, once you've got a PR that you think resolves the issue I'll run it across my app to ensure it hasn't cropped up again.

**Comment by xiphiaz** [comment]
@dwightwatson can you provide a stack trace of the error you get - I'm unable to replicate with my use case - though in my case the morph many isn't a lot of foreign models so I'm probably nowhere near a memory limit. 

A stack trace might help understand how the path differs from previously. I've added a new PR, with the original breaking code so you can copy that into your code to attempt to reproduce.

**Comment by dwightwatson** [comment]
The stack trace I got from production was actually truncated by Sentry, I will try to reproduce it in full tomorrow. In the meantime, here is the stack trace from running the query within `tinker`.

```
PHP Fatal error:  Allowed memory size of 536870912 bytes exhausted (tried to allocate 82 bytes) in /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Database/Connection.php on line 323
PHP Stack trace:
PHP   1. {main}() /home/vagrant/laravel/artisan:0
PHP   2. Illuminate\Foundation\Console\Kernel->handle() /home/vagrant/laravel/artisan:36
PHP   3. Symfony\Component\Console\Application->run() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php:107
PHP   4. Symfony\Component\Console\Application->doRun() /home/vagrant/laravel/vendor/symfony/console/Application.php:120
PHP   5. Symfony\Component\Console\Application->doRunCommand() /home/vagrant/laravel/vendor/symfony/console/Application.php:189
PHP   6. Illuminate\Console\Command->run() /home/vagrant/laravel/vendor/symfony/console/Application.php:838
PHP   7. Symfony\Component\Console\Command\Command->run() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Console/Command.php:136
PHP   8. Illuminate\Console\Command->execute() /home/vagrant/laravel/vendor/symfony/console/Command/Command.php:256
PHP   9. Illuminate\Container\Container->call() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Console/Command.php:150
PHP  10. call_user_func_array:{/home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Container/Container.php:503}() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Container/Container.php:503
PHP  11. Illuminate\Foundation\Console\TinkerCommand->fire() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Container/Container.php:503
PHP  12. Psy\Shell->run() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Foundation/Console/TinkerCommand.php:54
PHP  13. Symfony\Component\Console\Application->run() /home/vagrant/laravel/vendor/psy/psysh/src/Psy/Shell.php:283
PHP  14. Psy\Shell->doRun() /home/vagrant/laravel/vendor/symfony/console/Application.php:120
PHP  15. Psy\ExecutionLoop\ForkingLoop->run() /home/vagrant/laravel/vendor/psy/psysh/src/Psy/Shell.php:317
PHP  16. Psy\ExecutionLoop\Loop->run() /home/vagrant/laravel/vendor/psy/psysh/src/Psy/ExecutionLoop/ForkingLoop.php:79
PHP  17. Psy\ExecutionLoop\{closure}() /home/vagrant/laravel/vendor/psy/psysh/src/Psy/ExecutionLoop/Loop.php:128
PHP  18. eval() /home/vagrant/laravel/vendor/psy/psysh/src/Psy/ExecutionLoop/Loop.php:76
PHP  19. Illuminate\Database\Eloquent\Builder->get() /home/vagrant/laravel/vendor/psy/psysh/src/Psy/ExecutionLoop/Loop.php(76) : eval()'d code:1
PHP  20. Illuminate\Database\Eloquent\Builder->eagerLoadRelations() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php:174
PHP  21. Illuminate\Database\Eloquent\Builder->loadRelation() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php:429
PHP  22. Illuminate\Database\Eloquent\Relations\Relation->getEager() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php:460
PHP  23. Illuminate\Database\Eloquent\Relations\MorphTo->get() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Relations/Relation.php:113
PHP  24. Illuminate\Database\Eloquent\Relations\Relation->__call() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Relations/Relation.php:113
PHP  25. call_user_func_array:{/home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Relations/Relation.php:331}() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Relations/Relation.php:331
PHP  26. Illuminate\Database\Eloquent\Builder->get() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Relations/Relation.php:331
PHP  27. Illuminate\Database\Eloquent\Builder->getModels() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php:168
PHP  28. Illuminate\Database\Query\Builder->get() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php:409
PHP  29. Illuminate\Database\Query\Builder->runSelect() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php:1383
PHP  30. Illuminate\Database\Connection->select() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php:1406
PHP  31. Illuminate\Database\Connection->run() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Database/Connection.php:324
PHP  32. Illuminate\Database\Connection->runQueryCallback() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Database/Connection.php:611
PHP  33. Illuminate\Database\Connection->Illuminate\Database\{closure}() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Database/Connection.php:644
PHP  34. PDOStatement->fetchAll() /home/vagrant/laravel/vendor/laravel/framework/src/Illuminate/Database/Connection.php:323
```

**Comment by themsaid** [comment]
Hello everyone,
Can you let us know the status of this issue? Are you still experiencing the same memory problem in a recent release?

**Comment by themsaid** [comment]
Ping @dwightwatson

**Comment by dwightwatson** [comment]
Whoops, sorry - as per the above discussion this issue was fixed as far as I'm concerned in v5.1.26. Haven't run into a similar issue since.

**Comment by themsaid** [comment]
@dwightwatson thank you for updating us :)
